api/types/databaseservice.go

```diff
@@ -0,0 +1,111 @@
+/*
+Copyright 2022 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package types
+
+import (
+	"github.com/gravitational/trace"
+
+	"github.com/gravitational/teleport/api/utils"
+)
+
+// DatabaseService represents a DatabaseService (agent).
+type DatabaseService interface {
+	// ResourceWithLabels provides common resource methods.
+	ResourceWithLabels
+
+	// GetNamespace returns the resource namespace.
+	GetNamespace() string
+
+	// GetResourceMatchers returns the resource matchers of the DatabaseService.
+	GetResourceMatchers() []Labels
+	// SetResourceMatchers sets the resource matchers.
+	SetResourceMatchers([]Labels)
+}
+
+// NewDatabaseServiceV1 creates a new DatabaseService instance.
+func NewDatabaseServiceV1(meta Metadata, spec DatabaseServiceSpecV1) (*DatabaseServiceV1, error) {
+	s := &DatabaseServiceV1{
+		ResourceHeader: ResourceHeader{
+			Metadata: meta,
+		},
+		Spec: spec,
+	}
+
+	if err := s.CheckAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return s, nil
+}
+
+func (s *DatabaseServiceV1) setStaticFields() {
+	s.Kind = KindDatabaseService
+	s.Version = V1
+}
+
+// CheckAndSetDefaults checks and sets default values for any missing fields.
+func (s *DatabaseServiceV1) CheckAndSetDefaults() error {
+	s.setStaticFields()
+
+	return trace.Wrap(s.ResourceHeader.CheckAndSetDefaults())
+}
+
+// GetResourceMatchers returns the resource matchers of the DatabaseService.
+func (s *DatabaseServiceV1) GetResourceMatchers() []Labels {
+	return s.Spec.ResourceMatchers
+}
+
+// SetResourceMatchers sets the resource matchers.
+func (s *DatabaseServiceV1) SetResourceMatchers(resourceMatchers []Labels) {
+	s.Spec.ResourceMatchers = resourceMatchers
+}
+
+// GetNamespace returns the resource namespace.
+func (s *DatabaseServiceV1) GetNamespace() string {
```

Comment:
From this point onwards this all looks like boilerplate. Is there a particular reason we need those methods? It's hard to say just by looking at this PR, since all of this is unused.

Reply:
They exist to satisfy the type inference from `DatabaseServiceV1` to the interface `DatabaseService`
It's used in `GRPCServer.GetAllDatabaseServices` which I'm changing in the next commit to convert from `DatabaseService` to `DatabaseServiceV1`

Reply:
Ack, thanks for explaining.